File: app/src/main/java/com/vyy/sekerimremake/features/chart/presenter/master/ChartMasterFragment.java
package com.vyy.sekerimremake.features.chart.presenter.master;

import android.content.Context;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.AnimationUtils;
import android.view.animation.LayoutAnimationController;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.app.SharedElementCallback;
import androidx.fragment.app.Fragment;
import androidx.navigation.NavController;
import androidx.navigation.Navigation;
import androidx.navigation.fragment.FragmentNavigator;
import androidx.recyclerview.widget.LinearLayoutManager;

import com.vyy.sekerimremake.R;
import com.vyy.sekerimremake.features.chart.data.source.local.ChartDbHelper;
import com.vyy.sekerimremake.databinding.FragmentChartBinding;

import java.util.List;
import java.util.Map;

public class ChartMasterFragment extends Fragment implements ChartAdapter.OnRowClickListener {

    private FragmentChartBinding binding;
    private Context mContext;
    private NavController navController;
    private int scrollPosition;

    public ChartMasterFragment() {
        // Required empty public constructor
    }

    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(context);
        mContext = context;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        //To fill the chart with random data.
//        fillChartWithRandomData(0, 12, 1, 29);

    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        binding = FragmentChartBinding.inflate(inflater, container, false);
        scrollPosition = 0;
        if (savedInstanceState != null) {
            scrollPosition = savedInstanceState.getInt("scrollPosition");
        } else {
            if (getArguments() != null) {
                scrollPosition = getArguments().getInt("scrollPosition");
            }
        }
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        navController = Navigation.findNavController(view);

        ChartAdapter chartAdapter = new ChartAdapter(mContext, new ChartDbHelper(mContext)
                .getEveryone(), this);
        binding.recyclerViewChart.setAdapter(chartAdapter);
        LinearLayoutManager linearLayoutManager = new LinearLayoutManager(mContext);
        binding.recyclerViewChart.setLayoutManager(linearLayoutManager);
        LayoutAnimationController layoutAnimationController = AnimationUtils
                .loadLayoutAnimation(mContext, R.anim.recyclerview_layout_animation);
        binding.recyclerViewChart.setLayoutAnimation(layoutAnimationController);
        if (chartAdapter.getItemCount() > 1) {
            binding.recyclerViewChart.scrollToPosition(chartAdapter.getItemCount() - 1);
        }
        binding.floatingActionButtonAddRow.setOnClickListener(v -> {
            Bundle bundle = new Bundle();
            bundle.putBoolean("button_flag", true);
            navController.navigate(R.id.action_chartMasterFragment_to_chartDetailsFragment, bundle);
         });
    }

    @Override
    public void onRowClick(int position, final View view) {
        Bundle bundle = new Bundle();
        bundle.putBoolean("button_flag", false);
        bundle.putInt("position", position);

        final FragmentNavigator.Extras extras = new FragmentNavigator.Extras.Builder()
                .addSharedElement(view,
                        view.getTransitionName())
                .build();
        setEnterSharedElementCallback(new SharedElementCallback() {
            @Override
            public void onMapSharedElements(List<String> names, Map<String, View> sharedElements) {
                super.onMapSharedElements(names, sharedElements);
                sharedElements.put(view.getTransitionName(), view);
            }
        });
        navController.navigate(
                R.id.action_chartMasterFragment_to_chartDetailsFragment,
                bundle,
                null,
                extras);

    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putInt("scrollPosition", scrollPosition);
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        binding = null;
    }

    //To fill the chart with random data.

//    private void fillChartWithRandomData(int monthStart, int monthEnd, int dayStart, int dayEnd) {
//        ChartDbHelper helper = new ChartDbHelper(mContext);
//        ChartRowModel model;
//        for(int i = monthStart; i < monthEnd; i++) {
//            for(int j = dayStart; j < dayEnd; j++){
//                model = new ChartRowModel(
//                        -1,
//                        j,
//                        i,
//                        2021,
//                        new Random().nextInt(10) > 1 ? new Random().nextInt(75) + 65 : 0,
//                        new Random().nextInt(10) > 1 ? new Random().nextInt(95) + 95 : 0,
//                        new Random().nextInt(10) > 1 ? new Random().nextInt(75) + 65 : 0,
//                        new Random().nextInt(10) > 1 ? new Random().nextInt(95) + 95 : 0,
//                        new Random().nextInt(10) > 1 ? new Random().nextInt(75) + 65 : 0,
//                        new Random().nextInt(10) > 1 ? new Random().nextInt(95) + 95 : 0,
//                        new Random().nextInt(10) > 5 ? new Random().nextInt(75) + 85 : 0
//                );
//                helper.addOne(model);
//            }
//        }
//    }


}

File: app/src/main/java/com/vyy/sekerimremake/features/chart/presenter/master/ChartMasterFragment.kt
package com.vyy.sekerimremake.features.chart.presenter.master

import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.fragment.app.activityViewModels
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import androidx.navigation.fragment.FragmentNavigatorExtras
import androidx.navigation.fragment.findNavController
import androidx.recyclerview.widget.LinearLayoutManager
import com.vyy.sekerimremake.MainViewModel
import com.vyy.sekerimremake.R
import com.vyy.sekerimremake.databinding.FragmentChartBinding
import com.vyy.sekerimremake.features.chart.domain.model.ChartDayModel
import com.vyy.sekerimremake.features.settings.domain.model.UserModel
import com.vyy.sekerimremake.features.settings.presenter.UsersAdapter
import com.vyy.sekerimremake.utils.FirestoreConstants
import com.vyy.sekerimremake.utils.Response
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.launch
import java.util.*

@AndroidEntryPoint
class ChartMasterFragment : Fragment() {
    //TODO: BaseFragment for binding.
    private var _binding: FragmentChartBinding? = null
    private val binding get() = _binding!!

    private val viewModelMain: MainViewModel by activityViewModels()

    private var scrollPosition: Int = 0

    private val usersAdapter: UsersAdapter by lazy {
        UsersAdapter { position: Int ->
            onUserClicked(position)
        }
    }

    private val chartAdapter: ChartAdapter by lazy {
        ChartAdapter { chartDayModel, chartDayView ->
            onRowClick(chartDayModel, chartDayView)
        }
    }


    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?
    ): View {
        _binding = FragmentChartBinding.inflate(inflater, container, false)
        scrollPosition =
            savedInstanceState?.getInt("scrollPosition") ?: (arguments?.getInt("scrollPosition")
                ?: 0)

        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initUsersRecyclerView()
        initChartRecyclerView()
        setFloatingActionButton()

        viewLifecycleOwner.lifecycleScope.launch {
            viewLifecycleOwner.repeatOnLifecycle(Lifecycle.State.STARTED) {
                launch {
                    viewModelMain.userResponse.collectLatest { response ->
                        when (response) {
                            is Response.Success -> {
                                response.data?.let { data ->
                                    handleSelectedUser(data)
                                }
                            }
                            is Response.Error -> {
                                Log.e("MonitoredsFragment", response.message)
                            }
                            else -> {
                                // TODO
                            }
                        }
                    }
                }
                launch {
                    viewModelMain.chartResponse.collectLatest { response ->
                        when (response) {
                            is Response.Success -> {
                                binding.recyclerViewChart.apply {
                                    scheduleLayoutAnimation()
                                    chartAdapter.submitList(response.data)
                                    if (scrollPosition != 0) {
                                        scrollToPosition(scrollPosition)
                                    }
                                }
                            }
                            is Response.Error -> {
                                chartAdapter.submitList(null)
                                Log.d("ChartMasterFragment", response.message)
                            }
                            else -> {
                                chartAdapter.submitList(null)
                                //TODO
                            }
                        }
                    }
                }
            }
        }
    }

    private fun initUsersRecyclerView() {
        binding.recyclerViewUsers?.apply {
            adapter = usersAdapter
            setHasFixedSize(true)
        }
    }

    //TODO: Revisit
//TODO: Loading Dialog
    private fun initChartRecyclerView() {
        val linearLayoutManager = LinearLayoutManager(requireContext())
        linearLayoutManager.reverseLayout = true
        linearLayoutManager.stackFromEnd = true
        binding.recyclerViewChart.apply {
            adapter = chartAdapter
            layoutManager = linearLayoutManager
            setHasFixedSize(true)
        }
    }

    private fun setFloatingActionButton() {
        binding.floatingActionButtonAddRow.setOnClickListener {
            val chartResponse = viewModelMain.chartResponse.value
            if (chartResponse is Response.Success) {
                val calendar = Calendar.getInstance()
                val action =
                    ChartMasterFragmentDirections.actionChartMasterFragmentToChartDetailsFragment(
                        day = calendar[Calendar.DAY_OF_MONTH].toString(),
                        month = calendar[Calendar.MONTH].toString(),
                        year = calendar[Calendar.YEAR].toString()
                    )
                findNavController().navigate(action)
            }
        }
    }

    private fun handleSelectedUser(mUser: UserModel) {
        val userList = mutableListOf<UserModel>().apply {
            add(UserModel(uid = mUser.uid, name = getString(R.string.my_values)))
        }

        if (mUser.monitoreds.isNullOrEmpty()) {
            binding.recyclerViewUsers?.visibility = View.GONE
            usersAdapter.selectedPosition = -1
        } else {
            binding.recyclerViewUsers?.visibility = View.VISIBLE
            val monitoredList = mUser.monitoreds!!.filter { monitored ->
                monitored[FirestoreConstants.NAME] != null && monitored[FirestoreConstants.UID] != null
            }.map { monitored ->
                UserModel(
                    uid = monitored[FirestoreConstants.UID],
                    name = monitored[FirestoreConstants.NAME],
                    userName = monitored[FirestoreConstants.USER_NAME]
                )
            }
            userList.addAll(monitoredList)
            if (usersAdapter.selectedPosition < 0 || usersAdapter.selectedPosition >= userList.size) {
                usersAdapter.selectedPosition = 0
            }
        }

        if (usersAdapter.selectedPosition <= 0) {
            viewModelMain.getChart(mUser.uid!!)
            binding.constraintLayoutFloatActionGroup?.visibility = View.VISIBLE
        } else {
            userList[usersAdapter.selectedPosition].uid?.let {
                viewModelMain.getChart(it)
            }
            binding.constraintLayoutFloatActionGroup?.visibility = View.GONE
        }

        usersAdapter.submitList(userList)
    }

    private fun onUserClicked(position: Int) {
        val userResponse = viewModelMain.userResponse.value as? Response.Success ?: return
        val currentUser = usersAdapter.currentList.getOrNull(position) ?: return

        val uid = if (currentUser.uid.isNullOrEmpty() || userResponse.data?.uid == currentUser.uid) {
            usersAdapter.selectUser(0)
            binding.constraintLayoutFloatActionGroup?.visibility = View.VISIBLE
            userResponse.data?.uid
        } else {
            usersAdapter.selectUser(position)
            binding.constraintLayoutFloatActionGroup?.visibility = View.GONE
            currentUser.uid
        }

        uid?.let { viewModelMain.getChart(it) }
    }

    private fun onRowClick(day: ChartDayModel, view: View?) {
        if (view != null && view.transitionName != null) {
            val extras = FragmentNavigatorExtras(Pair(view, view.transitionName))

            setEnterSharedElementCallback(object : androidx.core.app.SharedElementCallback() {
                override fun onMapSharedElements(
                    names: List<String>, sharedElements: MutableMap<String, View>
                ) {
                    super.onMapSharedElements(names, sharedElements)
                    sharedElements[view.transitionName] = view
                }
            })
            val action =
                ChartMasterFragmentDirections.actionChartMasterFragmentToChartDetailsFragment(
                    day = day.dayOfMonth.toString(),
                    month = day.month.toString(),
                    year = day.year.toString()
                )
            findNavController().navigate(
                action, extras
            )
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putInt("scrollPosition", scrollPosition)
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }
}